test/test_device_radix_sort.cu

```diff
@@ -626,73 +628,218 @@ template <typename KeyT>
+        UnsignedBits inc_bits = 0;
+        RandomBits(inc_bits);
+        // twiddled_key < bits_max at this point.
+        UnsignedBits inc = std::min(1 + inc_bits % max_inc, max_bits - twiddled_key);
+        twiddled_key += inc;
+        
+        // Generate random run length (ensure there are enough values to fill the rest).
+        size_t run_bits = 0;
+        RandomBits(run_bits);
+        size_t run_length = std::min(1 + run_bits % max_run, num_items - i);
+        if (twiddled_key == max_bits) run_length = num_items - i;
+        size_t run_end = i + run_length;
+        
+        // Fill the array.
+        UnsignedBits key = TraitsT::TwiddleOut(twiddled_key);
+        for (; i < run_end; ++i)
+        {
+            h_key_bits[i] = key;
+        }
+    }
+}
+
+
 /**
  * Initialize solution
  */
-template <bool IS_DESCENDING, typename KeyT>
+template <bool IS_DESCENDING, typename KeyT, typename OffsetT>
 void InitializeSolution(
     KeyT    *h_keys,
-    int     num_items,
+    OffsetT num_items,
     int     num_segments,
-    int     *h_segment_offsets,
+    bool    pre_sorted,
+    OffsetT *h_segment_offsets,
     int     begin_bit,
     int     end_bit,
-    int     *&h_reference_ranks,
+    OffsetT *&h_reference_ranks,
     KeyT    *&h_reference_keys)
 {
-    typedef Pair<KeyT, int> PairT;
-
-    PairT *h_pairs = new PairT[num_items];
-
-    int num_bits = end_bit - begin_bit;
-    for (int i = 0; i < num_items; ++i)
+    if (pre_sorted)
     {
-
-        // Mask off unwanted portions
-        if (num_bits < static_cast<int>(sizeof(KeyT) * 8))
+        printf("Shuffling reference solution on CPU (%d segments)\n", num_segments);
+        // Copy to the reference solution.
+        h_reference_keys = new KeyT[num_items];
+        if (IS_DESCENDING)
         {
-            unsigned long long base = 0;
-            memcpy(&base, &h_keys[i], sizeof(KeyT));
-            base &= ((1ull << num_bits) - 1) << begin_bit;
-            memcpy(&h_pairs[i].key, &base, sizeof(KeyT));
+            // Copy in reverse.
+            for (OffsetT i = 0; i < num_items; ++i)
+            {
+                h_reference_keys[i] = h_keys[num_items - 1 - i];
+            }
+            // Copy back.
+            memcpy(h_keys, h_reference_keys, num_items * sizeof(KeyT));
         }
         else
         {
-            h_pairs[i].key = h_keys[i];
+            memcpy(h_reference_keys, h_keys, num_items * sizeof(KeyT));
         }
 
-        h_pairs[i].value = i;
+        // Summarize the pre-sorted array (element, 1st position, count).
+        struct Element
+        {
+            KeyT key;
+            size_t num;
+            size_t index;
+        };
+        std::vector<Element> summary;
+        KeyT cur_key = h_reference_keys[0];
+        summary.push_back(Element{cur_key, 1, 0});
+        for (size_t i = 1; i < num_items; ++i)
+        {
+            // TODO: handle -0.0, +0.0 and NaNs correctly.
```

Comment:
Will this affect correctness when testing floats? If so, we should add this before merging.

Reply:
Added permutation-based tests for floats.